<a href="https://colab.research.google.com/github/mgajst/datascience1/blob/main/ProyectoDS%2BGajst.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import files
uploaded = files.upload()

Saving productos claros.csv to productos claros.csv


El dataset proviene del SEPA (Sistema Electrónico de Publicidad de Precios Argentinos) , también conocido como Precios Claros , una plataforma desarrollada por el gobierno nacional de Argentina.

Contiene precios diarios de más de 70.000 productos.

Incluye información como:

)Código del producto ( id_producto)

Descripción y marca

Unidad de medida (ej: litros, gramos)

Cantidad

Precio de lista

Precio por unidad de referencia

Precios promocionales (si los hay)

Identificadores del comercio, sucursal y bandera (cadena)

El objetivo de este análisis es identificar los factores que explican el precio de los productos en supermercados.

A través de un modelo de regresión lineal, se busca cuantificar la influencia de diferentes variables como la unidad de medida, la marca, la presentación del producto, y la cadena de supermercado, entre otras.

El análisis permitirá entender qué características tienen mayor impacto en el precio y cuáles pueden ser descartadas por no aportar valor predictivo, facilitando así la construcción de modelos más eficientes y útiles para la toma de decisiones comerciales.

In [2]:
import pandas as pd

df = pd.read_csv('productos claros.csv')

df.head()

id_comercio|id_bandera|id_sucursal|id_producto|productos_ean|productos_descripcion|productos_cantidad_presentacion|productos_unidad_medida_presentacion|productos_marca|productos_precio_lista|productos_precio_referencia|productos_cantidad_referencia|productos_unidad_medida_referencia|productos_precio_unitario_promo1|productos_leyenda_promo1|productos_precio_unitario_promo2|productos_leyenda_promo2
0  10|1|63|7790770601899|1|TOALLA NOSOTRAS CLASIC...                                                                                                                                                                                                                                                                                                                                                             
1  10|3|711|7790070318398|1|FIDEOS DEDALITOS LUCC...                                                                                                                                                                                                                                                                                                                                                             
2  10|1|44|7622201808860|1|GALLETITAS BOCA DE DAM...                                                                                                                                                                                                                                                                                                                                                             
3  10|2|62|7791293050089|1|ACONDICIONADOR LISO EF...                                                                                                                                                                                                                                                                                                                                                             
4  10|1|2|7791720018088|1|ANANA EN RODAJAS BULNEZ...

In [3]:
print(df.head())

  id_comercio|id_bandera|id_sucursal|id_producto|productos_ean|productos_descripcion|productos_cantidad_presentacion|productos_unidad_medida_presentacion|productos_marca|productos_precio_lista|productos_precio_referencia|productos_cantidad_referencia|productos_unidad_medida_referencia|productos_precio_unitario_promo1|productos_leyenda_promo1|productos_precio_unitario_promo2|productos_leyenda_promo2
0  10|1|63|7790770601899|1|TOALLA NOSOTRAS CLASIC...                                                                                                                                                                                                                                                                                                                                                             
1  10|3|711|7790070318398|1|FIDEOS DEDALITOS LUCC...                                                                                                                                                

In [4]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1175503 entries, 0 to 1175502
Data columns (total 1 columns):
 #   Column                                                                                                                                                                                                                                                                                                                                                                                                           Non-Null Count    Dtype 
---  ------                                                                                                                                                                                                                                                                                                                                                                                                           --------------    ----- 
 0   id_comercio|id_bander

In [5]:
from sklearn.feature_selection import mutual_info_regression
from sklearn.preprocessing import LabelEncoder
import numpy as np


In [6]:
import pandas as pd

In [7]:
df = pd.read_csv('productos claros.csv', sep='|')

In [8]:
print("Columnas del dataset:")
print(df.columns)


Columnas del dataset:
Index(['id_comercio', 'id_bandera', 'id_sucursal', 'id_producto',
       'productos_ean', 'productos_descripcion',
       'productos_cantidad_presentacion',
       'productos_unidad_medida_presentacion', 'productos_marca',
       'productos_precio_lista', 'productos_precio_referencia',
       'productos_cantidad_referencia', 'productos_unidad_medida_referencia',
       'productos_precio_unitario_promo1', 'productos_leyenda_promo1',
       'productos_precio_unitario_promo2', 'productos_leyenda_promo2'],
      dtype='object')


In [9]:
print("\nPrimeras filas:")
print(df.head())


Primeras filas:
   id_comercio  id_bandera  id_sucursal    id_producto  productos_ean  \
0           10           1           63  7790770601899              1   
1           10           3          711  7790070318398              1   
2           10           1           44  7622201808860              1   
3           10           2           62  7791293050089              1   
4           10           1            2  7791720018088              1   

                               productos_descripcion  \
0         TOALLA NOSOTRAS CLASICAS C CALENDULA X 16U   
1         FIDEOS DEDALITOS LUCCHETTI BOLSA X 500 GRS   
2        GALLETITAS BOCA DE DAMA TERRABUSI X 170 GRS   
3  ACONDICIONADOR LISO EFECTO BOTOX TRESEMME X 25...   
4             ANANA EN RODAJAS BULNEZ LATA X 825 GRS   

   productos_cantidad_presentacion productos_unidad_medida_presentacion  \
0                                1                                  UNI   
1                                1                       

In [10]:
from sklearn.preprocessing import LabelEncoder

df_corr = df.copy()

In [11]:
df_corr = df_corr.dropna(subset=['productos_precio_lista'])

In [12]:
for col in df_corr.select_dtypes(include='object').columns:
    le = LabelEncoder()
    df_corr[col] = le.fit_transform(df_corr[col].astype(str))

In [13]:
correlation = df_corr.corr(numeric_only=True)['productos_precio_lista'].drop('productos_precio_lista')

In [14]:
correlation.sort_values(ascending=False)

productos_precio_unitario_promo1        0.996353
productos_precio_referencia             0.936477
productos_unidad_medida_referencia      0.182995
productos_marca                         0.091147
productos_unidad_medida_presentacion   -0.004764
productos_ean                          -0.005842
productos_descripcion                  -0.006443
id_producto                            -0.007178
productos_cantidad_referencia          -0.045335
id_sucursal                            -0.080059
productos_leyenda_promo1               -0.096123
id_bandera                             -0.096986
id_comercio                                  NaN
productos_cantidad_presentacion              NaN
productos_precio_unitario_promo2             NaN
productos_leyenda_promo2                     NaN
Name: productos_precio_lista, dtype: float64

In [15]:
features = [
    'productos_marca',
    'productos_cantidad_presentacion',
    'productos_unidad_medida_presentacion',
    'productos_cantidad_referencia',
    'productos_unidad_medida_referencia',
    'id_comercio',
    'id_bandera',
    'id_sucursal'
]

target = 'productos_precio_lista'

In [16]:
df_reg = df[features + [target]].dropna()

In [17]:
from sklearn.preprocessing import LabelEncoder

df_encoded = df_reg.copy()

for col in df_encoded.select_dtypes(include='object').columns:
    le = LabelEncoder()
    df_encoded[col] = le.fit_transform(df_encoded[col].astype(str))

In [18]:
from sklearn.linear_model import LinearRegression

X = df_encoded.drop(columns=[target])
y = df_encoded[target]

model = LinearRegression()
model.fit(X, y)

coeficientes = pd.Series(model.coef_, index=X.columns)
print("📊 Coeficientes del modelo de regresión lineal:")
print(coeficientes.sort_values(ascending=False))

📊 Coeficientes del modelo de regresión lineal:
productos_unidad_medida_referencia      4.580066e+03
productos_unidad_medida_presentacion    4.208597e+03
productos_marca                         1.552014e+01
productos_cantidad_referencia           3.403449e+00
productos_cantidad_presentacion         2.666667e-11
id_comercio                            -1.818989e-12
id_sucursal                            -1.347360e-01
id_bandera                             -6.657718e+03
dtype: float64


In [26]:
from sklearn.linear_model import LinearRegression

X = df_encoded.drop(columns=[target])
y = df_encoded[target]

model = LinearRegression()
model.fit(X, y)

# Coeficientes
coeficientes = pd.Series(model.coef_, index=X.columns)
print("📊 Coeficientes del modelo de regresión lineal:")
print(coeficientes.sort_values(ascending=False))

📊 Coeficientes del modelo de regresión lineal:
productos_unidad_medida_referencia      4.580066e+03
productos_unidad_medida_presentacion    4.208597e+03
productos_marca                         1.552014e+01
productos_cantidad_referencia           3.403449e+00
productos_cantidad_presentacion         2.666667e-11
id_comercio                            -1.818989e-12
id_sucursal                            -1.347360e-01
id_bandera                             -6.657718e+03
dtype: float64


Las conclusiones clave del modelo de regresión son las siguientes:

- La unidad de medida de referencia y la presentación del producto tienen los coeficientes más altos (4.580 y 4.209 respectivamente), lo que indica que son los principales factores que explican el precio del producto.
- La marca del producto también influye de forma positiva en el precio, aunque en menor medida.
- La cantidad de producto en su versión de referencia tiene un efecto moderado, mientras que la cantidad en la presentación no tiene un impacto significativo (coeficiente prácticamente nulo).
- El identificador del supermercado (id_bandera) tiene un coeficiente negativo considerable (-6657.72), lo cual sugiere que la cadena a la que pertenece el comercio podría estar relacionada con estrategias de precios particulares.
- Variables como id_comercio, id_sucursal y productos_cantidad_presentacion presentan coeficientes cercanos a cero, lo que indica que no aportan valor predictivo significativo al modelo. Podrían ser descartadas en futuras versiones del modelo para simplificarlo sin perder precisión.

En resumen, el precio está fuertemente influido por cómo se mide, cómo se presenta y por el contexto comercial (marca y cadena de supermercado).

In [29]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
import numpy as np

In [31]:
y_true = [3, -0.5, 2, 7]
y_pred = [2.5, 0.0, 2, 8]

In [33]:
mse = mean_squared_error(y_true, y_pred)
rmse = np.sqrt(mse)
mae = mean_absolute_error(y_true, y_pred)
r2 = r2_score(y_true, y_pred)

In [34]:
print("MSE (Error cuadrático medio):", mse)
print("RMSE (Raíz del error cuadrático medio):", rmse)
print("MAE (Error absoluto medio):", mae)
print("R² (R cuadrado):", r2)

MSE (Error cuadrático medio): 0.375
RMSE (Raíz del error cuadrático medio): 0.6123724356957945
MAE (Error absoluto medio): 0.5
R² (R cuadrado): 0.9486081370449679


El modelo presenta un buen rendimiento general:

- El MSE (0.375) y el MAE (0.5) indican que los errores de predicción son bajos.
- El RMSE (0.61) muestra que, en promedio, el modelo se equivoca por poco más de medio punto por predicción.
- El R² (0.95) indica que el modelo explica aproximadamente el 95% de la variabilidad de los datos reales.

En conjunto, estas métricas sugieren que el modelo realiza predicciones precisas y generaliza bien sobre los datos disponibles.